modules/stub/src/stub.js

```diff
@@ -1,241 +1,282 @@
+      if (event && event.data) {
 
-                    /**
-                     * If it came in as a string it will be return in the same
-                     * manner
-                     */
+        if (typeof event.data === 'string') {
 
-                    returnMsg = JSON.stringify(returnMsg);
+          try {
 
-                  }
+            /**
+             * Try to parse the data from the event; there is no guarantee
+             * that this string is a stringified object – in fact it's
+             * probably more likely that it's not.
+             */
 
-                  event.source.postMessage(returnMsg, TARGET_ORIGIN);
+            event.data = JSON.parse(event.data);
 
-                };
+          } catch (ignore) {/* oh well... */}
 
-                win[TCF_API_NAME](
-                  command,
-                  version,
-                  wrapperCallback,
-                  parameter,
-                );
+        }
 
-              }
+        if (typeof event.data === 'object') {
 
-            }
+          if (isLeader && event.data[TCF_POST_MSG_CALL]) {
 
-          };
+            /**
+             * Since this instance is at the top, then the CMP exists here
+             * and we'll wrap all postMessage callbacks. Otherwise, this
+             * message should be ignored because this instance isn't tasked
+             * with handling it.
+             */
 
-          win.addEventListener(EVENT_NAME, postMessageEventHandler);
+            const {command, version, callId, parameter} = event.data[TCF_POST_MSG_CALL];
 
-        } else {
+            const wrapperCallback = (returnValue, success) => {
 
-          let callId = 0;
-          const callMap = new Map();
+              if (event.source && event.source.postMessage) {
 
-          win[TCF_API_NAME] = (command, version, callback, parameter) => {
+                event.source.postMessage({
+                  [TCF_POST_MSG_RETURN]: {
+                    returnValue,
+                    success,
+                    callId,
+                  },
+                }, TARGET_ORIGIN);
 
-            callMap.set(callId, callback);
+              }
 
-            const msg = {
-              [TCF_POST_MSG_CALL]: {
-                command,
-                parameter,
-                version,
-                callId,
-              },
             };
 
-            win.parent.postMessage(msg, TARGET_ORIGIN);
-            callId++;
+            win[TCF_API_NAME](
+              command,
+              version,
+              wrapperCallback,
+              parameter,
+            );
 
-          };
+          } else if (!isLeader && event.data[TCF_POST_MSG_RETURN]) {
 
-          const responseHandler = () => {
+            const {callId, returnValue, success} = event.data[TCF_POST_MSG_RETURN];
 
-            let json;
+            if (callMap.has(callId)) {
 
-            if (typeof event.data === 'string') {
+              const call = callMap.get(callId);
```

Comment:
I would suggest adding some error handling to ensure that call isn't null or undefined.  We recently encountered an issue working with a different CMP where the other CMP returned multiple times.  This was for the __cmp('getConsentData', ...) command which doesn't apply here.  But it highlights the possibility that some bug in another CMP might result in callbacks to commands (other than addEventListener) being called multiple times.